### PR TITLE
introduce Region kind and type constructors

### PR DESCRIPTION
**Pull request by mlutze**
related to #3832

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/fmt/SimpleType.scala [hunk 1]
```diff
@@ -244,6 +244,11 @@ object SimpleType {
     */
   case class Tuple(fields: List[SimpleType]) extends SimpleType
 
+  /**
+    * A region-to-boolean conversion type.
+    */
+  case object RegionToBool extends SimpleType
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/fmt/SimpleType.scala:250]: I think RegionToBool and RegionToStar should be kept right after each other.

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala [hunk 2]
```diff
@@ -335,13 +335,27 @@ object TypeConstructor {
...
+      * The shape of a region term is Region[l].
       */
-    def kind: Kind = Kind.Bool ->: Kind.Star
+    def kind: Kind = Kind.Region ->: Kind.Star
+  }
+
+  /**
+    * A type constructor that represents the a region.
+    */
+  case class Region(sym: Symbol.RegionSym) extends TypeConstructor {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:350]: Maybe this goesl before RegionToStar.

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala [hunk 3]
```diff
@@ -335,13 +335,27 @@ object TypeConstructor {
   }
 
   /**
-    * A type constructor that represent the type of regions.
+    * A type constructor that represents the type of term-level regions.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:338]: I would avoid using the word term.

File: main/src/ca/uwaterloo/flix/api/lsp/Indexer.scala [hunk 4]
```diff
@@ -195,7 +195,7 @@ object Indexer {
       Index.occurrenceOf(exp0)
 
     case Expression.Scope(sym, _, exp, _, _, loc) =>
-      val tpe = Type.mkRegion(sym.tvar.ascribedWith(Kind.Bool), loc)
+      val tpe = Type.mkRegionTerm(sym.tvar.ascribedWith(Kind.Bool), loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/Indexer.scala:198]: RegionToStar

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala [hunk 5]
```diff
@@ -613,7 +613,9 @@ object SemanticTokensProvider {
     case TypeConstructor.Intersection => false
     case TypeConstructor.Difference => false
     case TypeConstructor.Effect(_) => false
-    case TypeConstructor.Region => false
+    case TypeConstructor.RegionToStar => false
+    case TypeConstructor.Region(_) => false
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:617]: Swap order.

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala [hunk 6]
```diff
@@ -335,13 +335,27 @@ object TypeConstructor {
...
+    def kind: Kind = Kind.Region ->: Kind.Bool
+  }
+
+  /**
+    * A type constructor converting Region to Star.
+    */
+  case object RegionToStar extends TypeConstructor {
     /**
-      * The shape of a region is Region[l].
+      * The shape of a region term is Region[l].
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:356]: not term

**Comment by mlutze** [comment]
Closing per discussion

**Pull request closed without merging**